### PR TITLE
[SPARK-23611][SQL] Add a helper function to check exception for expr evaluation

### PR DESCRIPTION
**Pull request by maropu**
## What changes were proposed in this pull request?
This pr added a helper function in `ExpressionEvalHelper` to check exceptions in all the path of expression evaluation.

## How was this patch tested?
Modified the existing tests.

### CODE REVIEW

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala [hunk 1]
```diff
@@ -93,6 +95,40 @@ trait ExpressionEvalHelper extends GeneratorDrivenPropertyChecks {
...
+      inputRow: InternalRow,
+      expectedErrMsg: String)(
+      implicit classTag: ClassTag[T]): Unit = {
+    val clazz = classTag.runtimeClass
+    val serializer = new JavaSerializer(new SparkConf()).newInstance
+    val resolver = ResolveTimeZone(new SQLConf)
+    val expr = resolver.resolveTimeZones(
+      serializer.deserialize[Expression](serializer.serialize(expression)))
+
+    def checkException(eval: => Unit, testMode: String): Unit = {
```
Comment (hvanhovell) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala]: Can we use the existing `intercept[...]` infrastructure here (not sure if it is possible)?

Comment (maropu) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala]: ok, I'll reconsider again. thanks.

Comment (hvanhovell) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala]: Thanks!

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala [hunk 2]
```diff
@@ -93,6 +95,40 @@ trait ExpressionEvalHelper extends GeneratorDrivenPropertyChecks {
     }
   }
 
+  protected def checkExceptionInExpression[T <: Throwable](
+      expression: => Expression,
+      inputRow: InternalRow,
+      expectedErrMsg: String)(
+      implicit classTag: ClassTag[T]): Unit = {
+    val clazz = classTag.runtimeClass
+    val serializer = new JavaSerializer(new SparkConf()).newInstance
```
Comment (hvanhovell) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala]: This code is also present in `checkEvaluation` can you factor it out into a common method?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala [hunk 3]
```diff
@@ -133,37 +169,39 @@ trait ExpressionEvalHelper extends GeneratorDrivenPropertyChecks {
     }
   }
 
+  private def evaluateWithoutCodegen(
```
Comment (hvanhovell) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala]: Why are we adding this method?

Comment (maropu) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala]: Just for readability (so, reverting is ok to me).

Comment (hvanhovell) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala]: Why not rename the actual method?

Comment (maropu) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala]: aha, ok. I'll fix that way. thanks.

**Comment by maropu** [comment]
@hvanhovell ok, check again? Thanks!

**Comment by hvanhovell** [comment]
LGTM - merging to master!

**Pull request closed without merging**